### PR TITLE
FOUNDATIONS: Remove Intern

### PR DESCRIPTION
**Pull request by ADILATIC**
Closes: #49

### CODE REVIEW

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by glhays** [comment]
@ADILATIC please put issue closing in you comment at the top of this PR.
The format has to be as follows to work.
Closes: #issuenumber

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

**Comment by ADILATIC** [comment]
image

File: InternTrack.Core.Api/Services/Foundations/Interns/InternService.cs [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 using System;
 using System.Linq;
+using System.Net.Sockets;
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api/Services/Foundations/Interns/InternService.cs:8]: I am not sure why it is being used. I hope this is not incorrect using statement.

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+                broker.SelectInternByIdAsync(It.IsAny<Guid>()))
+                    .ThrowsAsync(sqlException);
+
+            //when
+            ValueTask<Intern> removeInternTask =
+                this.internService.RemoveInternByIdAsync(someInternId);
+
+            //then
+            await Assert.ThrowsAsync<InternDependencyException>(() =>
+                removeInternTask.AsTask());
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs:46]: This is old version of how we handle exception. Move to new version.

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectInternByIdAsync(It.IsAny<Guid>()),
+                    Times.Once());
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs:50]: Times.Once() -> Times.Once
Remove brackets

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+                    expectedInterndependencyException))),
+                        Times.Once());
+
+            this.storageBrokerMock.VerifyNoOtherCalls();
+            this.loggingBrokerMock.VerifyNoOtherCalls();
+            this.dateTimeBrokerMock.VerifyNoOtherCalls();
+        }
+
+        [Fact]
+        public async Task ShouldThrowdependencyExceptionOnRemoveByIdIfDatabaseUpdateExceptionOccursAndLogItAsync()
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs:63]: This should come after db update concurrency exception

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+                broker.SelectInternByIdAsync(It.IsAny<Guid>()))
+                    .ThrowsAsync(databaseUpdateException);
+
+            // when 
+            ValueTask<Intern> removeInternTask =
+                this.internService.RemoveInternByIdAsync(someInternId);
+
+            //then
+            await Assert.ThrowsAsync<InternDependencyException>(() =>
+                removeInternTask.AsTask());
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs:85]: Old version here. Move to new

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Logic.RemoveById.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                await this.internService.RemoveInternByIdAsync(inputInternId);
+
+            // then
+            actualIntern.Should().BeEquivalentTo(expectedIntern);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectInternByIdAsync(inputInternId),
+                    Times.Once());
+            
+            this.storageBrokerMock.Verify(broker =>
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Logic.RemoveById.cs:53]: No need for brackets ()'

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.RemoveById.cs [hunk 7]
```diff
@@ -0,0 +1,109 @@
+﻿// ---------------------------------------------------------------
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.RemoveById.cs:1]: Change the order of vno in your logic test as well. I could not leave comment there.

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.RemoveById.cs [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                .BeEquivalentTo(expectedInternValidationException);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.SelectInternByIdAsync(It.IsAny<Guid>()),
+                    Times.Once());
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionsAs(
+                    expectedInternValidationException))),
+                        Times.Once());
```
Comment (ElbekDeveloper) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.RemoveById.cs:98]: No need for bracket ()

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.RemoveById.cs (hunk 7 above)
Comment (ADILATIC) [reply InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.RemoveById.cs:1]: What does VNO stand for?

Comment (ADILATIC) [reply InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Validations.RemoveById.cs:1]: Neverminded figured it out.

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                    .ThrowsAsync(sqlException);
+
+            //when
+            ValueTask<Intern> removeInternAsync =
+                this.internService.RemoveInternByIdAsync(someInternId);
+
+            InternDependencyException actualInternDependencyException =
+                await Assert.ThrowsAsync<InternDependencyException>(
+                    removeInternAsync.AsTask);
+
```
Comment (hassanhabib) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs:45]: Remove extra line

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+using Microsoft.Data.SqlClient;
+using Microsoft.EntityFrameworkCore;
+using Moq;
+using Xunit;
+
+namespace InternTrack.Core.Api.Tests.Unit.Services.Foundations.Interns
+{
+    public partial class InternServiceTests
+    {
+
```
Comment (hassanhabib) [inline InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs:20]: Remove extra line here

File: InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs (hunk 5 above)
Comment (hassanhabib) [reply InternTrack.Core.Api.Tests.Unit/Services/Foundations/Interns/InternServiceTests.Exceptions.RemoveById.cs:91]: @ADILATIC this is the old version of handling exceptions like @ElbekDeveloper said - move to the new pattern - check other similar code for examples.

**Review by glhays — APPROVED** [review]

**Pull request merged**